cli/module_registry.go

```diff
@@ -380,6 +399,67 @@ func getNextModuleUploadRequest(file *os.File) (*apppb.UploadModuleFileRequest,
 	}, nil
 }
 
+func validateModuleFile(client *viamClient, moduleID moduleID, tarballPath, version string) error {
+	getModuleResp, err := client.getModule(moduleID)
+	if err != nil {
+		return err
+	}
+	entrypoint, err := getEntrypointForVersion(getModuleResp.Module, version)
+	if err != nil {
+		return err
+	}
+	//nolint:gosec
+	file, err := os.Open(tarballPath)
+	if err != nil {
+		return err
+	}
+	// TODO(APP-2226): support .tar.xz
+	if !strings.HasSuffix(file.Name(), ".tar.gz") {
+		return errors.New("you must upload your module in the form of a .tar.gz")
+	}
+	archive, err := gzip.NewReader(file)
+	if err != nil {
+		return err
+	}
+
+	tarReader := tar.NewReader(archive)
+	filesWithSameNameAsEntrypoint := []string{}
+	for {
+		if err := client.c.Context.Err(); err != nil {
+			return err
+		}
+		header, err := tarReader.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return errors.Wrapf(err, "error reading %s", file.Name())
+		}
+		path := header.Name
+
+		// if path == entrypoint, we have found the right file
+		if filepath.Clean(path) == filepath.Clean(entrypoint) {
+			info := header.FileInfo()
+			if info.Mode().Perm()&0o100 == 0 {
+				return errors.Errorf(
+					"the provided tarball %q contained a file at the entrypoint %q, but that file is not marked as executable",
+					tarballPath, entrypoint)
+			}
+			// executable file at entrypoint. validation succeeded.
+			return nil
+		}
+		if filepath.Base(path) == filepath.Base(entrypoint) {
```

Comment:
> can we not just use the base check first as they should have the same entrypoint if that is true?

All of this logic is to catch people who have `entrypoint: ./main` but there is a file called `./realdir/main` so we want to tell people that they have a bad entrypoint but also let them know that there is a suspiciously similarly named file in the tarball.

>  why do filepath.base() after filepath.clean() 

Because base just checks for filename but clean removes the `.` and `..` characters so they can be compared properly

Reply:
but can't we just do path.clean(path.base()) at the top level to check that? instead of having separate checks?